src/imp/mbedtls.rs

```diff
@@ -0,0 +1,546 @@
+                Box::from_raw(self.ca_certs);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MidHandshakeTlsStream<S> {
+    stream: TlsStream<S>,
+    error: Error,
+}
+
+pub enum HandshakeError<S> {
+    Failure(Error),
+    WouldBlock(MidHandshakeTlsStream<S>),
+}
+
+impl<S> MidHandshakeTlsStream<S> {
+    pub fn get_ref(&self) -> &S {
+        self.stream.get_ref()
+    }
+
+    pub fn get_mut(&mut self) -> &mut S {
+        self.stream.get_mut()
+    }
+}
+
+impl<S> MidHandshakeTlsStream<S>
+where
+    S: io::Read + io::Write,
+{
+    pub fn handshake(self) -> Result<TlsStream<S>, HandshakeError<S>> {
+        Ok(self.stream)
+    }
+}
+
+#[derive(Clone)]
+pub struct TlsConnector {
+    min_protocol: Option<Protocol>,
+    max_protocol: Option<Protocol>,
+    root_certificates: Vec<::Certificate>,
+    identity: Option<::Identity>,
+    accept_invalid_certs: bool,
+    accept_invalid_hostnames: bool,
+    use_sni: bool,
+}
+
+impl TlsConnector {
+    pub fn new(builder: &TlsConnectorBuilder) -> Result<TlsConnector, Error> {
+        let trust_roots = if builder.root_certificates.len() > 0 {
+            builder.root_certificates.clone()
+        } else {
+            load_system_trust_roots()?
+        };
+
+        Ok(TlsConnector {
+            min_protocol: builder.min_protocol,
+            max_protocol: builder.max_protocol,
+            root_certificates: trust_roots,
+            identity: builder.identity.clone(),
+            accept_invalid_certs: builder.accept_invalid_certs,
+            accept_invalid_hostnames: builder.accept_invalid_hostnames,
+            use_sni: builder.use_sni
+        })
+    }
+
+    pub fn connect<S>(&self, domain: &str, stream: S) -> Result<TlsStream<S>, HandshakeError<S>>
+    where
+        S: io::Read + io::Write
+    {
+        // If any of the ? fail then memory leaks ...
+
+        let identity = if let Some(identity) = &self.identity {
+            let mut keys = (identity.0).0.private_keys().collect::<Vec<_>>();
+            let certificates = (identity.0).0.certificates().collect::<Vec<_>>();
+
+            if keys.len() != 1 {
+                return Err(HandshakeError::Failure(Error::Custom("Unexpected number of keys in PKCS12 file".to_owned())))
+            }
+            if certificates.len() == 0 {
+                return Err(HandshakeError::Failure(Error::Custom("PKCS12 file is missing certificate chain".to_owned())))
+            }
+
+            let mut cert_chain = vec![];
+            for cert in certificates {
+                cert_chain.push(cert.0?);
+            }
+
+            fn pk_clone(pk: &mut Pk) -> TlsResult<Pk> {
+                let der = pk.write_private_der_vec()?;
+                Pk::from_private_key(&der, None)
+            }
+            let key = Box::new(keys.pop().unwrap().0.map_err(|_| TlsError::PkInvalidAlg)?);
+
+            Some((cert_chain, key))
+        } else {
+            None
+        };
+
+        unsafe {
```

Comment:
I'm not sure I'm on board with this much unsafe code for just establishing a TLS connection. Why can't the mbedtls crate provide a safe interface?

Reply:
The current Rust API for MbedTLS doesn't take ownership of anything. So if you run your entire server on a single stack, no unsafe is needed. However, if you want to do something else, you need self-referential structs.